### DISCUSSION TITLE
Caret Position Variable

### DISCUSSION THREAD

**Discussion by vrajur** [Taitava/obsidian-shellcommands#114] [Ideas - done & released]
Hey there!

Thanks for making this awesome plugin, I use it all the time! I wanted to suggest adding a variable accessing the cursor position (line and column number), which is helpful for passing to some commands I tend to run. I wanted to post here first to get your thoughts, but I already have a local branch that I could push if you're interested in taking a look.

Thanks

**Comment by Taitava** [comment]
Hi! 🙂 

Thank you for your suggestion! I think it's a good idea! A few points on my mind:
1. I think we could use the term _caret_ instead of _cursor_, as I understand that _cursor_ usually refers to mouse cursor, while _caret_ refers to the blinking text input pointer. It's easy to mix them up 🙂 . [Here's at least one article describing the difference between _cursor_ and _caret_](https://en.wikibooks.org/wiki/Windows_Programming/Interfacing#Cursor_Vs_Caret).
2. What would be the variable's name? I'm considering these, but you can suggest others too:
    - `{{caret_position}}`, or
    - Simply `{{caret}}`, but only if there's no future use for any other properties a caret might have?
3. How to tell which value is wanted, _line_ or _column_?
    - Using a parameter to tell which value to get: `{{caret:line}}` / `{{caret:column}}`, or
    - Something else?
4. Have you seen other plugins that access the caret position? I'm especially interested about plugins that have a variable system, and might have already a good name for a caret position variable, but maybe there is no such plugin. But even if there's a plugin that accesses the caret position without having it as a variable, I'd like to know.
5. Yes, a pull request is welcome after these decisions have been made. 👍 

Can you also provide an example command that could be inserted to [this documentation page](https://publish.obsidian.md/shellcommands/Example+shell+commands/Example+shell+commands) to demonstrate the new variable, please? You can either:
- Fork the [documentation repository](https://github.com/Taitava/obsidian-shellcommands-documentation) and add it yourself (create a new note in _Example commands_ folder and add a link to the new note in _Example commands.md_ ). Before doing this, we should decide the variable name and how it's used. Then make a PR to that repo and I'll merge and publish the documentation change when releasing a version that contains the new variable.
- Just post the example command here in this discussion and I'll add it to the documentation. Can be done anytime, I can edit the variable name if needed.

And thank you for your nice feedback! 🌞

**Reply by vrajur** [reply]
Awesome! Let me respond to your points:

1. Thanks for pointing this out - yeah let's go with `caret` as you indicated
2. I would think `{{caret_position}}` for the name. I wouldn't rule out having a more general variable for `caret` down the line, which `caret_position` could get refactored into (if it that bridge ever gets crossed). 
3. Yeah, I was thinking it could follow your current syntax for specifying parameter names such that:

| Variable | Description | 
| --- | --- | 
|  `{{cursor_position}}` | Returns both line and column (e.g. `1442:10`) |
| `{{cursor_position:line}}` | Returns just line (e.g. `1442`) |
| `{{cursor_position:column}}` | Returns just column (e.g. `10`) | 

4. I haven't really looked to closely for other examples, but with a quick search I found the [CursorJumper](https://github.com/SilentVoid13/Templater/blob/master/src/editor/CursorJumper.ts) class for the [Templater](https://github.com/SilentVoid13/Templater) plugin. From what I understand the terminology should be caret there as well, but this class seems to control where the caret jumps to when using an obsidian template containing the [`tp.file.cursor(order?: number)`](https://silentvoid13.github.io/Templater/docs/internal-variables-functions/internal-modules/file-module/#documentation) function.
5. I'll push up a working branch at `#114-caret-position-variable` to iterate on the implementation then if that's ok. I'll wait to open a PR until we've decided things here. 

I can also add an example shell command to your docs repo once a the code PR is up ready for review.

Thanks for your quick feedback on my request!

**Reply by vrajur** [reply]
Actually, I'm not able to push a branch (getting a permission denied error on pushing). It might be something silly on my end, but I have a branch locally that you could take a look at so let me know how you'd go from here :slightly_smiling_face:

**Reply by FelipeRearden** [reply]
Great idea @vrajur !!!

Hello @Taitava !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!

Sorry for my absence these past two weeks (We talk more about that in other post....)

I saw this an my mind start to generate ideas !!!!!

> Have you seen other plugins that access the caret position? I'm especially interested about plugins that have a variable system, and might have already a good name for a caret position variable, but maybe there is no such plugin. But even if there's a plugin that accesses the caret position without having it as a variable, I'd like to know.

Obsidian Go to Line https://github.com/phibr0/obsidian-go-to-line

Would be amazing if we have the same syntax between the two plugins which is `line:column` 

I'm pretty excited about this command `Go To Line with Clipboard Content`

https://github.com/phibr0/obsidian-go-to-line#2-go-to-line-with-clipboard-contents


> Can you also provide an example command that could be inserted to this documentation page to demonstrate the new variable, please? You can either:

My first ideia would be an SC to copy  the current caret caret position (line+column) to the clipboard: 
- Would be amazing to create template notes that receives content from other Obsidian Notes and Clipboard 
- Or change the caret position after inserting content in a note (under an specific header for example)

Step 1: 	SC =`echo {{caret}}` [output`Clipboard`]  
Step 2: 	Use `Go To Line with Clipboard Content` 


Second idea:  Apply Patterns plugin changes the caret position after `find/replace` . This way I could save the caret position before perform an `find/replace` action then get back. More here https://github.com/jglev/obsidian-apply-patterns-plugin/discussions/31

Step 1: 	SC =`echo {{caret}}` [output`Clipboard`]  
Step 2:   Apply Pattern `find/replace` To the whole document  
Step 3: 	Use `Go To Line with Clipboard Content` 


I let you guys know when I found another user case

### Two Things to Think in the future about `caret`

1- New Output `Current File: Caret Position line:column`
2- New variables `{{caret_content:line}}` and  `{{caret_content:line:column}}` that get the word or the full line based on the position inserted by the user in the variable

Let’s keep talking :)

Have a great day!

**Reply by Taitava** [reply]
> 2. I would think `{{caret_position}}` for the name. I wouldn't rule out having a more general variable for `caret` down the line, which `caret_position` could get refactored into (if it that bridge ever gets crossed).

Good, let's use `{{caret_position}}` then 🙂. @FelipeRearden posted another variable idea `{{caret_content}}`, so if we'll want to implement that too in the future, both of the names share the same logic.

> 
> 3. Yeah, I was thinking it could follow your current syntax for specifying parameter names such that:
> 
> | Variable | Description |
> | ---- | --- |
> | `{{cursor_position}}` |	Returns both line and column (e.g. `1442:10`) |
> | `{{cursor_position:line}}` |	Returns just line (e.g. `1442`) |
> | `{{cursor_position:column}}` |	Returns just column (e.g. `10`) |

Yea, let's do all of those three. `:` is a good default separator, and if something else is wanted, a longer style `{{caret_position:line}}, {{caret_position:column}}` can be used, where `, ` is a freely decideable separator.

> 5. I'll push up a working branch at `#114-caret-position-variable` to iterate on the implementation then if that's ok. I'll wait to open a PR until we've decided things here.

That's a good branch name 🙂. There are two possible branches that you can use as a starting point:
 - `main` - it's a bit old, but not too much. I tend to keep `main` very close to a state that contains the latest published version (+ some changes to files like `README.md` and `CHANGELOG.md`), not so much in-development stuff, but I might consider changing this habit later.
 - `0.8.0` - in development atm. I just pushed this branch yesterday so maybe you have already based your work on `main`, and it's ok. This newer stuff contains some changes to how variables' help texts are managed, due to #33 . But whatever your work is based on, I can handle merging. When you create a PR, the target branch can be `main`, I'll probably change it to `0.9.0` when I create that branch after `0.8.0` is out. In other words, I'm planning this new variable to be released in version `0.9.0`.

> 
> 
> I can also add an example shell command to your docs repo once a the code PR is up ready for review.

Thank you! 🙂 

> 
> Thanks for your quick feedback on my request!

You are welcome!


> Actually, I'm not able to push a branch (getting a permission denied error on pushing). It might be something silly on my end, but I have a branch locally that you could take a look at so let me know how you'd go from here 🙂

I guess you cloned this repository directly? You need to fork it first, then GitHub will let you push to your own fork, and then you can create a PR from your fork repository to my repository. 🙂

**Reply by Taitava** [reply]
> Hello @Taitava !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
> 
> Sorry for my absence these past two weeks (We talk more about that in other post....)

Hello! 🙂 No problem, take your time! 😉 

> 
> I saw this an my mind start to generate ideas !!!!!
> 
> > Have you seen other plugins that access the caret position? I'm especially interested about plugins that have a variable system, and might have already a good name for a caret position variable, but maybe there is no such plugin. But even if there's a plugin that accesses the caret position without having it as a variable, I'd like to know.
> 
> Obsidian Go to Line https://github.com/phibr0/obsidian-go-to-line

That looks handy! 🙂 Someday I need to create a vault of plugins that I should explore.

> 
> Would be amazing if we have the same syntax between the two plugins which is `line:column`

Yes, we'll have it! 😎 

> My first ideia would be an SC to copy the current caret caret position (line+column) to the clipboard:
> 
>  * Would be amazing to create template notes that receives content from other Obsidian Notes and Clipboard

Sorry, I'm not quite understanding what you mean by this idea? 🙂

> 
>  * Or change the caret position after inserting content in a note (under an specific header for example)
> 
> Step 1: SC =`echo {{caret}}` [output`Clipboard`] Step 2: Use `Go To Line with Clipboard Content`

Yea, although you can use _Go to line_ also for doing the first step (with _Copy cursor line number_ command). But if you have a shell command that searches for text in your notes, then outputs the line and column, you can store the output to clipboard like you do in your example, and then use `Go To Line with Clipboard Content`. In that case, you'd not need {{caret_position}} at all, though.


> 
> I let you guys know when I found another user case
> ### Two Things to Think in the future about `caret`
> 
> 1- New Output `Current File: Caret Position line:column`

Do you mean that in the output settings, there would be a form field for inputting specific line and column numbers where the output should be made? You can open a new idea discussion for this and answer there 🙂 . 

> 2- New variables `{{caret_content:line}}` and `{{caret_content:line:column}}` that get the word or the full line based on the position inserted by the user in the variable

Yea, I need to think about all possible variations, word, line, paragraph, chapter, or everything under the current heading (with/without the heading), etc. You can create a new discussion for this too, and answer there. Good ideas! 🙂 

> 
> Let’s keep talking :)
> 
> Have a great day!

Nice day to you too! 🙂

**Reply by FelipeRearden** [reply]
> Sorry, I'm not quite understanding what you mean by this idea? 🙂

I have a lot of workflows to insert content to Obsidian from external sources (web,Books,Podcast apps,iOS ...). 

This way I have to set a note template for each one of this sources and insert this external content in the right place in this template. 

Having more than one SC to manipulate the caret position, I can manipulate the caret position before insert content.

> Yea, although you can use Go to line also for doing the first step (with Copy cursor line number command). But if you have a shell command that searches for text in your notes, then outputs the line and column, you can store the output to clipboard like you do in your example, and then use Go To Line with Clipboard Content. In that case, you'd not need {{caret_position}} at all, though.

The difference between SC and the Go to Line (Copy cursor line number command) is that we can define a specific caret position with SC instead of only have the clipboard of the current caret position as a source of action ! 

> In that case, you'd not need {{caret_position}} at all, though.

For one workflow I will have 3/4 SC with different caret positions in the note instead of just using one from the Go To Line. That's why this new variable is a killer feature !!! We will have a way to navigate between the note without typing anything !!!!!!  

> Do you mean that in the output settings, there would be a form field for inputting specific line and column numbers where the output should be made?

Yes !!!!! `echo {{selection}} [output=Current File: Caret Position:16:03]` line 16 column 3

With this new output we close the matrix of possibilities to `Current File` top/bottom/caret_position/line:column !!!!! Users could insert file in every place possible in a note 

> You can open a new idea discussion for this and answer there 🙂 .

I will thanks :)

> Yea, I need to think about all possible variations, word, line, paragraph, chapter, or everything under the current heading (with/without the heading), etc. 

Yeah, I know I went far from the initial idea but my mind was blowing with ideas related to this topic :)

It's a lot of to think about :)
You can create a new discussion for this too, and answer there. Good ideas! 🙂

I will thanks @Taitava !!!!

**Reply by vrajur** [reply]
Hi @FelipeRearden! I didn't realize there were so many possibilities for this simple addition - thanks for sharing!

@Taitava thanks for the feedback and yeah you were right I cloned rather than forked (told you something silly on my end :stuck_out_tongue:). But anyways, I rebased my branch onto your `0.8.0` updates and opened a PR against that rather than `main` just to keep the diffs small. Happy to change this to `main` or another target if you'd like though. I'd appreciate your input whenever you have a chance. In the meantime, I will fork the docs repo and add updates there as well.

Thanks!

**Reply by Taitava** [reply]
> I rebased my branch onto your `0.8.0` updates and opened a PR against that rather than `main` just to keep the diffs small.

Nice! And it's good idea to keep the diff small by selecting `0.8.0` as the target. I'll change it later when it's time. I'll comment more on the PR.

Thank you! 🙂

**Reply by FelipeRearden** [reply]
> Hi @FelipeRearden! I didn't realize there were so many possibilities for this simple addition - thanks for sharing!

Thanks @vrajur !!!!!

**Comment by Taitava** [comment]
# Created issue #119 for this
I totally forgot that usually I have a habit to create an issue for implementation.

Normal workflow:
1. Have a discussion (like we've had here).
2. Create an issue that contains a summary of all **decisions** from the discussions.
3. Branch name should have the issue's #id number (now the branch has this discussion's #id number, but its ok now).
4. Commits should also have the issue's #id number, because the the branch name will not persist in git log.

So the number 2 is something I forgot to do, but I did it now afterwards.
